src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,241 @@
+    functionName: 'approve',
+    args: [allowanceTarget as Address, amountToApprove],
+  })
+
+  const approveTx: TransactionRequestCIP42 & { amountToApprove: bigint } = {
+    from: from as Address,
+    to: sellTokenAddress as Address,
+    data,
+    type: 'cip42',
+    amountToApprove,
+  }
+
+  return approveTx
+}
+
+function createBaseSwapTransaction(unvalidatedSwapTransaction: SwapTransaction) {
+  const { from, to, value, data, gas } = unvalidatedSwapTransaction
+
+  const swapTx: TransactionRequestCIP42 & { gas: bigint } = {
+    from: from as Address,
+    to: to as Address,
+    value: BigInt(value ?? 0),
+    data: data as Hex,
+    // This may not be entirely accurate for now
+    // See https://www.notion.so/valora-inc/Fee-currency-selection-logic-4c207244893748bd85e23b754334f42d?pvs=4#8b7c27d31ebf4fca981f81e9411f86ee
+    // We'll try to improve this in our API
+    gas: BigInt(gas),
+    type: 'cip42',
+  }
+
+  return swapTx
+}
+
+function getMaxGasCost(txs: TransactionRequestCIP42[]) {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return maxGasCost
+}
+
+async function prepareTransactions(
+  fromToken: TokenBalanceWithAddress,
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction,
+  price: string,
+  feeCurrencies: TokenBalance[]
+): Promise<PreparedTransactionsResult> {
+  const baseApproveTx = createBaseApproveTransaction(updatedField, unvalidatedSwapTransaction)
+  const baseSwapTx = createBaseSwapTransaction(unvalidatedSwapTransaction)
+
+  const maxGasCosts: Array<{ feeCurrency: TokenBalance; maxGasCostInDecimal: BigNumber }> = []
+  for (const feeCurrency of feeCurrencies) {
+    if (feeCurrency.balance.isLessThanOrEqualTo(0)) {
+      // No balance, try next fee currency
+      continue
+    }
+
+    const feeCurrencyAddress = !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
+    const { maxFeePerGas, maxPriorityFeePerGas } = await estimateFeesPerGas(
+      publicClient.celo,
+      feeCurrencyAddress
+    )
+
+    const approveTx = {
+      ...baseApproveTx,
+      maxFeePerGas,
+      maxPriorityFeePerGas,
+      feeCurrency: feeCurrencyAddress,
+    }
+
+    // TODO maybe cache this? and add static padding when using non-native fee currency
+    try {
+      const approveGas = await publicClient.celo.estimateGas({
+        ...approveTx,
+        account: approveTx.from,
+      })
+      approveTx.gas = approveGas
+    } catch (e) {
+      if (e instanceof EstimateGasExecutionError) {
+        // Likely too much gas was needed
+        Logger.warn(
+          'SwapScreen@useSwapQuote',
+          `Couldn't estimate gas for approval with feeCurrency ${feeCurrency.symbol} (${feeCurrency.tokenId}), trying next feeCurrency`,
+          e
+        )
+        continue
+      }
+      throw e
```

Comment:
any reason why we want to throw here? i thought continuing to the next fee currency would be better than exiting from getting the quote entirely

Reply:
Yes this isn't supposed to happen. Trying the next fee currency isn't the appropriate action.